Add duplicate way removal as a post conflate op

bwitham opened the pull request:
This will (slightly) help clean up ref conflate output at intersections that don't conflate very well and end up leaving duplicate roads on top of each other.

Unrelated: Also made the `count` command output a little easier to read.

bwitham commented:
@curranMapper RE: your ref conflate duplicated road issue from Monday... I have three recommendations for you when using the Unifying alg with roads and ref conflate (one related to this PR):

1) Increase your road search radius in adv opts - Increasing the search radius from 15m to 30m got me cleaner merges using ref conflate with the Unifying alg. This is a little counter-intuitive b/c some of the roads aren't 30m apart but it does work. Your runtime will increase, of course.

2) Set the "Duplicate Way Removal strict tag matching" option to false in adv opts - I added duplicate way removal to the ops that run after conflate finishes as part of this PR. Setting that option will allow roads which are geometrically identical but have differing tags to be removed by the duplicate way remover post conflate.

3) Use Network alg instead of Unifying alg with roads - Using ref conflate with the Network alg instead of Unifying I got cleaner road merges with no dupe overlapping roads...no other changes were required.

Full disclosure, I did not look at all 291 road segments you flagged in the original output...only a handful of them.

sisskind commented:
@bwitham for your guidance above, is that documented for the average user?

bwitham commented:
I would classify this type of documentation under conflate "Conflate Best Practices" or "Conflate Tradecraft". hoot doesn't have much like that in the documentation right now, but I will add a section in somewhere. I know @curranMapper has been keeping documentation like that for years now. What I said above should still be verified as completely accurate. I only had time to test my findings on a half a dozen roads in that dataset. Its also worth noting what's said above may not apply to every road dataset you conflate...that's where the "Tradecraft" of conflation comes in. Knowing how to configure conflate options to get the best output if the default options don't give you desirable output.

curranMapper commented:
I would like to test and get you guys some feedback.

I tried option 3, (Network conflation) if this results in fixing the noted issue (still need to verify), the drawback to this options is that there are 585 conflict reviews to work through compared to 150 when running Unify.

I think if option 2 has solid results, It may be the most seamless improvement to user.  User wouldn't need to change any of the default settings, and issue might be remedied.  This would keep the number of reviews lower as well.  I need to test.

will report back here after further testing.

curranMapper commented:
**OPTION 3.**  Network conflation fixes the ‘overlapping highways’ issue, but requires user to walk through 530 reviews. The upload included some orphan nodes.

to compare that to Unify conflation (default settings), this results in 160 reviews, with 133 of these ‘overlapping way’ instances to fix in JOSM.  (290 fixes total vs the 530 reviews when running network conflation above)

**OPTION 2**: Unify conflation, setting the adv option 'Duplicate Way Removal strict tag matching` to false reduces the ‘overlapping highways’ issue from 133 to 19. (180 fixes total).  However, in this result, it opts to use the highway classification of the perpendicular road that is generally classified differently then the road that it is attached too.  (depicted in graphics below)



![image](https://user-images.githubusercontent.com/15927720/76088384-f1c91e00-5f85-11ea-8b70-515ab2de8ca1.png)

![image](https://user-images.githubusercontent.com/15927720/76088400-fdb4e000-5f85-11ea-9f35-5abb5c34aa65.png)


Is there a way these small segments can get automatically assigned the same highway classification of the larger feature it is now apart of?  That came to mind as one way to further remedy this issue, maybe another way comes to mind for you @bwitham ?

From workflow perspective, focus on the above is keeping it as seamless as possible for user, ie decrease need for users to alter from the default settings, or increase number of reviews needed to be worked through in the process.  Right now,  standard workflow to use unify (default), and then do further cleaning in JOSM still seems to be best option from that perspective.

bwitham commented:
Possibly if either `DuplicateWayRemover` is set to run after `WayJoinerOp` during post conflate instead OR `DuplicateWayRemover` is set to run both before and after `WayJoinerOp`, this problem may be solved. Will create an issue.

bwitham commented:
https://github.com/ngageoint/hootenanny/issues/3885

bwitham commented:
> **OPTION 3.** Network conflation fixes the ‘overlapping highways’ issue, but requires user to walk through 530 reviews. The upload included some orphan nodes.
> 
> to compare that to Unify conflation (default settings), this results in 160 reviews, with 133 of these ‘overlapping way’ instances to fix in JOSM. (290 fixes total vs the 530 reviews when running network conflation above) It is less work for the user to go the unify route.
> 
> **OPTION 2**: Unify conflation, setting the adv option 'Duplicate Way Removal strict tag matching` to false reduces the ‘overlapping highways’ issue from 133 to 19. (180 fixes total). However, in this result, it opts to use the highway classification of the perpendicular road that is generally classified differently then the road that it is attached too. (depicted in graphics below)
> 
> ![image](https://user-images.githubusercontent.com/15927720/76088384-f1c91e00-5f85-11ea-8b70-515ab2de8ca1.png)
> 
> ![image](https://user-images.githubusercontent.com/15927720/76088400-fdb4e000-5f85-11ea-9f35-5abb5c34aa65.png)
> 
> Is there a way these small segments can get automatically assigned the same highway classification of the larger feature it is now apart of? That came to mind as one way to further remedy this issue, maybe another way comes to mind for you @bwitham ?
> 
> From workflow perspective, focus on the above is keeping it as seamless as possible for user, ie decrease need for users to alter from the default settings, or increase number of reviews needed to be worked through in the process. Right now, standard workflow to use unify (default), and then do further cleaning in JOSM still seems to remain the best option with this one from that perspective.

@curranMapper I'm working on the fix for the problem described above. When you have a chance, can you refresh my memory on how you ran conflate for this? The results I'm getting are a lot different than the screenshots above, but it may just be I'm not running the conflate job the same way that you did. thanks.

curranMapper commented:
@bwitham 

Applies to both:
Ran a reference conflation type,
Used the ecopia dataset as reference layer, and OSM data as secondary layer.  
Changed the "attribute reference layer" to OSM, secondary layer.

then, OPTION 3.   changed the road engine from "Unify" to "Network"
![image](https://user-images.githubusercontent.com/15927720/77350105-689a4100-6d12-11ea-97ba-addf928f9d9d.png)

then OPTION 2.  Unchecked the "duplicate way removal strict matching" in Advanced options, and left all else as defaults.

![image](https://user-images.githubusercontent.com/15927720/77350386-d9415d80-6d12-11ea-8536-2f6216ada5ca.png)

Does that help?  There a different configuration I should be using?
Goal is to bring OSM/NOME secondary data highway tags and additional features to the Reference Ecopia dataset.

bwitham commented:
That's what I needed thanks. I was missing the step to change the attribute reference layer so wasn't seeing any of the OSM tags in the output. Now, I should be able to fix the tag issue with the duplicated ways.

bwitham commented:
So, it looks like I fixed the road classification problem you identified when I was working on something else after this was merged. The output looks good to me now, and there is only one overlapping highway in the output left. I added a test for the screenshot you posted above in #3927.
